storage/posix/antispam/badger.go

```diff
@@ -0,0 +1,416 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package badger provides a Tessera persistent antispam driver based on
+// BadgerDB (https://github.com/hypermodeinc/badger), a high-performance
+// pure-go DB with KV support.
+package badger
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/dgraph-io/badger/v4"
+	tessera "github.com/transparency-dev/trillian-tessera"
+	"github.com/transparency-dev/trillian-tessera/api/layout"
+	"github.com/transparency-dev/trillian-tessera/internal/otel"
+	"k8s.io/klog/v2"
+)
+
+const (
+	DefaultMaxBatchSize      = 1500
+	DefaultPushbackThreshold = 2048
+)
+
+var (
+	nextKey = []byte("@nextIdx")
+)
+
+// AntispamOpts allows configuration of some tunable options.
+type AntispamOpts struct {
+	// MaxBatchSize is the largest number of mutations permitted in a single BatchWrite operation when
+	// updating the antispam index.
+	//
+	// Larger batches can enable (up to a point) higher throughput, but care should be taken not to
+	// overload the Spanner instance.
+	//
+	// During testing, we've found that 1500 appears to offer maximum throughput when using Spanner instances
+	// with 300 or more PU. Smaller deployments (e.g. 100 PU) will likely perform better with smaller batch
+	// sizes of around 64.
+	MaxBatchSize uint
+
+	// PushbackThreshold allows configuration of when to start responding to Add requests with pushback due to
+	// the antispam follower falling too far behind.
+	//
+	// When the antispam follower is at least this many entries behind the size of the locally integrated tree,
+	// the antispam decorator will return tessera.ErrPushback for every Add request.
+	PushbackThreshold uint
+}
+
+// NewAntispam returns an antispam driver which uses Badger to maintain a mapping between
+// previously seen entries and their assigned indices.
+//
+// Note that the storage for this mapping is entirely separate and unconnected to the storage used for
+// maintaining the Merkle tree.
+//
+// This functionality is experimental!
+func NewAntispam(ctx context.Context, badgerPath string, opts AntispamOpts) (*AntispamStorage, error) {
+	if opts.MaxBatchSize == 0 {
+		opts.MaxBatchSize = DefaultMaxBatchSize
+	}
+	if opts.PushbackThreshold == 0 {
+		opts.PushbackThreshold = DefaultPushbackThreshold
+	}
+
+	// Open the Badger database located in the /tmp/badger directory.
+	// It is created if it doesn't exist.
+	db, err := badger.Open(badger.DefaultOptions(badgerPath))
+	if err != nil {
+		return nil, fmt.Errorf("failed to open badger: %v", err)
+	}
+
+	r := &AntispamStorage{
+		opts: opts,
+		db:   db,
+	}
+
+	go func() {
+		ticker := time.NewTicker(1 * time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+		again:
+			err := db.RunValueLogGC(0.7)
+			if err == nil {
+				goto again
```

Comment:
Wow, there's a name I haven't heard in a long time.

Reply:
https://www.youtube.com/watch?v=7SM_eJKVlzs&list=RD7SM_eJKVlzs#t=72